jmolecules-architecture/jmolecules-cqrs-architecture/src/main/java/org/jmolecules/architecture/cqrs/annotation/CommandHandler.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jmolecules.architecture.cqrs.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Identifies a command handler in the context of CQRS, i.e. logic to process a {@link Command}.
+ * The command handler may or may not reject the command. In case of processing, the handler
+ * takes care of orchestrating the business logic related to the command.
+ *
+ * @author Christian Stettler
+ * @author Henning Schwentner
+ * @author Stephan Pirnbaum
+ * @author Martin Schimak
+ * @author Oliver Drotbohm
+ * @see <a href="http://cqrs.files.wordpress.com/2010/11/cqrs_documents.pdf">CQRS Documents by Greg Young - Commands</a>
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
```

Comment:
Just noticed the target for these annotations. I think this one can be placed on a constructor as well. I don't think we should exclude that possibility at this level.

Also, being able to place this as meta-annotations would be very powerful. That way, frameworks like Axon could put these annotations on their own.

Reply:
Agree. Annotations to be used on a method should also be usable on constructors. Also, they should the the `ElementType.ANNOTATION_TYPE` target for use as meta-annotations.

Reply:
As meta-annotations is a topic relevant to jmolecules and not only this PR, let's tackle this with issue #40 

What would be the use case/benefit of a Handler annotation on a constructor and what should be the meaning of this? Will this make the complete class a handler? All methods contained in it? Or is the consturctor then a handler (which I have strong aversion to)?

As far as I'm concerned, you also cannot put a @ExceptionHandler or @GetMapping on a constructor.

Reply:
> As meta-annotations is a topic relevant to jmolecules and not only this PR, let's tackle this with issue #40

I don't think there's anything to do with the currently existing annotations as they already work as meta-annotations. See my comment on #40.

Regarding the constructor use: good point. @abuijze – can you elaborate?

Reply:
Axon allows developers to put annotations directly on the Command Model objects. An `@CommandHandler` annotation on a constructor means that the Framework is expected to create a new instance of an Aggregate when that command arrives. It passes the command as a parameter to the constructor.

There are some limitations, however. Since the return value of a constructor is the created object, and it's generally a very bad idea to return the entire model object, we extract the identifier of the created aggregate and return that as response to the command. Another limitation is that applying events on the instance isn't possible until after the constructor has executed, since there is no instance before then. Mainly because of that, we have stated supporting creation of aggregate on regular instance methods as well.

> Or is the consturctor then a handler (which I have strong aversion to)
Yes, that is actually what it means (in Axon). While I understand the aversion, we also need to realize that different people will have different opinions on what's best or good.